core/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -81,6 +95,141 @@ boolean recoverFromStore(final IndexShard indexShard, final boolean indexShouldE
         return false;
     }
 
+    boolean recoverFromLocalShards(BiConsumer<String, MappingMetaData> mappingUpdateConsumer, final IndexShard indexShard, final List<LocalShardSnapshot> shards) throws IOException {
+        if (canRecover(indexShard)) {
+            assert indexShard.recoveryState().getType() == RecoveryState.Type.LOCAL_SHARDS : "invalid recovery type: " + indexShard.recoveryState().getType();
+            if (indexShard.routingEntry().restoreSource() != null) {
+                throw new IllegalStateException("can't recover - restore source is not null");
+            }
+            if (shards.isEmpty()) {
+                throw new IllegalArgumentException("shards must not be empty");
+            }
+            Set<Index> indices = shards.stream().map((s) -> s.getIndex()).collect(Collectors.toSet());
+            if (indices.size() > 1) {
+                throw new IllegalArgumentException("can't add shards from more than one index");
+            }
+            for (ObjectObjectCursor<String, MappingMetaData> mapping : shards.get(0).getMappings()) {
+                mappingUpdateConsumer.accept(mapping.key, mapping.value);
+            }
+            for (ObjectObjectCursor<String, MappingMetaData> mapping : shards.get(0).getMappings()) {
+                indexShard.mapperService().merge(mapping.key,mapping.value.source(), MapperService.MergeReason.MAPPING_RECOVERY, true);
+            }
+            return executeRecovery(indexShard, () -> {
+                logger.debug("starting recovery from local shards {}", shards);
+                try {
+                    /*
+                     * TODO: once we upgraded to Lucene 6.1 use HardlinkCopyDirectoryWrapper to enable hardlinks if possible and enable it
```

Comment:
we don't do this anymore since all kinds tests will fail then... I can write a test that fails instead